neps/nep-0364.md

```diff
@@ -0,0 +1,171 @@
+# transaction id AKFwCuQ8g2a8t9SX7vcve348BvwoU3M42MSef7gy1kPA
+```
+
+With `iterations = 130` **all these calls return ExecutionError**: `'Exceeded the maximum amount of gas allowed to burn per contract.'`
+With iterations = 50 these are the results:
+
+```
+ed25518: tx id 6DcJYfkp9fGxDGtQLZ2m6PEDBwKHXpk7Lf5VgDYLi9vB (299 Tgas)
+schnorrkel: tx id ACYJh7YC4pQM8fu7DsjmmkSb3WSvikfQkVyej3htiuDQ (304 Tgas)
+ecdsa: tx id 2FhtFEbakuwQyHWcyeRoPkCDgXz4erTYSD75QkxDTb2e (185 Tgas)
+```
+
+- Performance in wall clock time when you compile the signature validation library directly from rust to native.
+  Here are the results on a AMD Ryzen 9 5900X 12-Core Processor machine:
+
+```
+# 10k signature verifications
+ed25519: took 387ms
+ecdsa: took 324ms
+shnorrkel: took 411ms
+```
+
+- Performance in wall clock time when you compile the library into wasm first and then use the single-pass compiler in Wasmer 1 to then compile to native.
+
+```
+ed25519: took 9926ms
+ecdsa: took 6725ms
+shnorrkel: took 10220ms
+```
+
+As an extra datapoint, when passing `--enable-simd` instead of `--singlepass`
+
+```
+ed25519: took 3085ms
+ecdsa: took 1895ms
+shnorrkel: took 3179ms
+```
+
+Steps to reproduce:
+commit: `31cf97fb2e155d238308f062c4b92bae716ac19f` in `https://github.com/blasrodri/near-test`
+
+```sh
+# wasi singlepass
+cargo wasi build --bin benches --release
+wasmer compile --singlepass ./target/wasm32-wasi/release/benches.wasi.wasm -o benches_singlepass
+wasmer run ./benches_singlepass
+```
+
+```sh
+# rust native
+cargo run --bin benches --release
+```
+
+Overall: the difference between the two versions (native vs wasi + singlepass are)
+
+```
+ed25519: 25.64x slower
+ecdsa: 20.75x slower
+shnorrkel: 24.86x slower
+```
+
+### What is the impact of not doing this?
+
+Costs of running IBC compatible trustless bridges would be very high. Plus, the fact that signature verification
+is such an expensive operation will force the contract to split the process of batch verification of signatures
+into multiple transactions.
+
+### Why is this design the best in the space of possible designs?
+
+Adding exisiting proved and vetted crypto crates into the runtime is a safe workaround. It will boost performance
+between 20-25x according to our benchmarks. This will both reduce operation costs significantly and will also
+enable the contract to verify all the signatures in one transaction, which will simplify the contract design.
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+Adding exisiting proved and vetted crypto crates into the runtime is a safe workaround. It will boost performance
+between 20-25x according to our benchmarks. This will both reduce operation costs significantly and will also
+enable the contract to verify all the signatures in one transaction, which will simplify the contract design.
```

Comment:
This is just a copy of the text from a subsection above, isn’t it?

One alternative I can think of is improving the runtime implementation sufficiently that it compiles WASM code to a sufficiently fast machine code in the first place. It won’t be as fast as LLVM produced code due to the need for gas accounting and such, and it would take a ton of effort, compared to adding a host function, but it is an alternative worth spelling out.

Another alternative is to deal with high cost of computing/verifying these signatures in some other manner – such as decreasing the overall cost of gas and increasing the limits of gas available to attach to the contract. Again, not as straightforward as a host function to implement and is less localized of a change.